datanode/broker/sqlstore_broker.go

```diff
@@ -265,7 +265,8 @@ func (b *SQLStoreBroker) processBlock(ctx context.Context, dbContext context.Con
 				beginBlock := e.(entities.BeginBlockEvent)
 				return entities.BlockFromBeginBlock(beginBlock)
 			case events.CoreSnapshotEvent:
-				b.snapshotTaken = true
+				// if a snapshot is taken on a protocol upgrade block, we want it to be taken synchronously as part of handling of protocol upgrade
+				b.snapshotTaken = !e.StreamMessage().GetCoreSnapshotEvent().ProtocolUpgradeBlock
```

Comment:
Wondering if this is necessary, the snapshots are sequential so if a snapshot is in progress, the second snapshot will not be allowed to proceed before the first has completed.  Once the first has completed, if the second is a request for exactly the same span then the snapshot process returns `ErrSnapshotExists`  - so maybe just need to handle this error in the protocol_upgrade_handler instead?